File: app/src/main/java/com/bhashasetu/app/database/LessonDao.java
package com.bhashasetu.app.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.bhashasetu.app.model.Lesson;
import com.bhashasetu.app.data.relation.LessonWithWords;
import com.bhashasetu.app.model.LessonWord;
import java.util.List;

@Dao
public interface LessonDao {

    // ✅ Basic Lesson operations
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertLesson(Lesson lesson);
    
    @Update
    void updateLesson(Lesson lesson);
    
    @Delete
    void deleteLesson(Lesson lesson);
    
    @Query("SELECT * FROM lessons ORDER BY category, orderInCategory ASC")
    LiveData<List<Lesson>> getAllLessons();

    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    LiveData<Lesson> getLessonById(long lessonId);
    
    @Query("SELECT * FROM lessons WHERE categoryId = :categoryId AND isActive = 1 ORDER BY 'order' ASC")
    LiveData<List<Lesson>> getLessonsByCategory(long categoryId);

    @Query("SELECT * FROM lessons WHERE difficultyLevel = :level AND isActive = 1 ORDER BY `order` ASC")
    LiveData<List<Lesson>> getLessonsByDifficulty(int level);

    @Query("SELECT * FROM lessons WHERE isActive = 1 ORDER BY `order` ASC")
    LiveData<List<Lesson>> getAllActiveLessons();

    @Query("UPDATE lessons SET totalWords = :wordCount WHERE id = :lessonId")
    void updateLessonWordCount(long lessonId, int wordCount);
    
    @Query("SELECT * FROM lessons WHERE level = :level ORDER BY category, orderInCategory ASC")
    LiveData<List<Lesson>> getLessonsByLevel(String level);
    
    @Query("SELECT * FROM lessons WHERE hasCompleted = 0 ORDER BY level, orderInCategory ASC LIMIT 1")
    LiveData<Lesson> getNextIncompleteLesson();
    
    @Query("SELECT DISTINCT category FROM lessons ORDER BY category ASC")
    LiveData<List<String>> getAllCategories();
    
    @Query("SELECT * FROM lessons WHERE id IN " +
           "(SELECT lessonId FROM lesson_words WHERE wordId = :wordId)")
    LiveData<List<Lesson>> getLessonsContainingWord(int wordId);
    
    @Insert
    void insertLessonWord(LessonWord lessonWord);
    
    @Query("DELETE FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId")
    void removeLessonWord(int lessonId, int wordId);
    
    @Transaction
    @Query("SELECT EXISTS(SELECT 1 FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId)")
    boolean lessonContainsWord(int lessonId, int wordId);
}

File: app/src/main/java/com/bhashasetu/app/data/dao/LessonDao.kt
package com.bhashasetu.app.data.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import androidx.room.Update
import com.bhashasetu.app.model.Lesson
import com.bhashasetu.app.data.relation.LessonWithWords
import com.bhashasetu.app.data.model.LessonWord

@Dao
interface LessonDao {

    // ✅ Basic Lesson operations
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertLesson(lesson: Lesson): Long

    @Update
    suspend fun updateLesson(lesson: Lesson)

    @Delete
    suspend fun deleteLesson(lesson: Lesson)

    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    fun getLessonById(lessonId: Long): LiveData<Lesson>

    @Query("SELECT * FROM lessons WHERE isActive = 1 ORDER BY `order` ASC")
    fun getAllActiveLessons(): LiveData<List<Lesson>>

    @Query("SELECT * FROM lessons WHERE categoryId = :categoryId AND isActive = 1 ORDER BY `order` ASC")
    fun getLessonsByCategory(categoryId: Long): LiveData<List<Lesson>>

    // ✅ FIXED: LessonWithWords queries using correct entity references

    /**
     * Get lesson with words - FIXED to use correct Word entity
     */
    @Transaction
    @Query("SELECT * FROM lessons WHERE id = :lessonId")
    fun getLessonWithWords(lessonId: Long): LiveData<LessonWithWords>

    /**
     * Get all lessons with their words - FIXED column references
     */
    @Transaction
    @Query("SELECT * FROM lessons WHERE isActive = 1 ORDER BY `order` ASC")
    fun getAllLessonsWithWords(): LiveData<List<LessonWithWords>>

    /**
     * Get lessons by category with words - FIXED entity references
     */
    @Transaction
    @Query("SELECT * FROM lessons WHERE categoryId = :categoryId AND isActive = 1 ORDER BY `order` ASC")
    fun getLessonsByCategoryWithWords(categoryId: Long): LiveData<List<LessonWithWords>>

    /**
     * ✅ FIXED: Complex query joining lessons and words with correct column names
     */
    @Transaction
    @Query("""
        SELECT l.* FROM lessons l
        INNER JOIN lesson_words lw ON l.id = lw.lessonId
        INNER JOIN vocabulary_words w ON lw.wordId = w.id
        WHERE w.difficulty = :difficulty 
        AND l.isActive = 1 
        AND w.isActive = 1
        GROUP BY l.id
        ORDER BY l.`order` ASC
    """)
    fun getLessonsWithWordsByDifficulty(difficulty: Int): LiveData<List<LessonWithWords>>

    /**
     * ✅ FIXED: Get lessons containing specific English word
     */
    @Transaction
    @Query("""
        SELECT l.* FROM lessons l
        INNER JOIN lesson_words lw ON l.id = lw.lessonId
        INNER JOIN vocabulary_words w ON lw.wordId = w.id
        WHERE w.englishWord LIKE '%' || :englishWord || '%'
        AND l.isActive = 1 
        AND w.isActive = 1
        GROUP BY l.id
        ORDER BY l.`order` ASC
    """)
    fun getLessonsContainingEnglishWord(englishWord: String): LiveData<List<LessonWithWords>>

    /**
     * ✅ FIXED: Get lessons containing specific Hindi word
     */
    @Transaction
    @Query("""
        SELECT l.* FROM lessons l
        INNER JOIN lesson_words lw ON l.id = lw.lessonId
        INNER JOIN vocabulary_words w ON lw.wordId = w.id
        WHERE w.hindiWord LIKE '%' || :hindiWord || '%'
        AND l.isActive = 1 
        AND w.isActive = 1
        GROUP BY l.id
        ORDER BY l.`order` ASC
    """)
    fun getLessonsContainingHindiWord(hindiWord: String): LiveData<List<LessonWithWords>>

    /**
     * ✅ NEW: Get lessons with quiz words only
     */
    @Transaction
    @Query("""
        SELECT l.* FROM lessons l
        INNER JOIN lesson_words lw ON l.id = lw.lessonId
        WHERE lw.includeInQuiz = 1
        AND l.isActive = 1
        GROUP BY l.id
        HAVING COUNT(lw.wordId) > 0
        ORDER BY l.`order` ASC
    """)
    fun getLessonsWithQuizWords(): LiveData<List<LessonWithWords>>

    // ✅ LessonWord junction table operations
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertLessonWord(lessonWord: LessonWord): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertLessonWords(lessonWords: List<LessonWord>)

    @Update
    suspend fun updateLessonWord(lessonWord: LessonWord)

    @Delete
    suspend fun deleteLessonWord(lessonWord: LessonWord)

    @Query("DELETE FROM lesson_words WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun removeLessonWord(lessonId: Long, wordId: Long)

    @Query("SELECT * FROM lesson_words WHERE lessonId = :lessonId ORDER BY orderInLesson ASC")
    fun getLessonWords(lessonId: Long): LiveData<List<LessonWord>>

    @Query("UPDATE lesson_words SET orderInLesson = :order WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun updateWordOrder(lessonId: Long, wordId: Long, order: Int)

    @Query("UPDATE lesson_words SET isKeyword = :isKeyword WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun updateKeywordStatus(lessonId: Long, wordId: Long, isKeyword: Boolean)

    @Query("UPDATE lesson_words SET includeInQuiz = :includeInQuiz WHERE lessonId = :lessonId AND wordId = :wordId")
    suspend fun updateQuizInclusion(lessonId: Long, wordId: Long, includeInQuiz: Boolean)
}